### PR TITLE
backport: Some wallet related PRs

### PR DESCRIPTION
**Pull request by xdustinface**
This PR:
- Includes backports related to the `unloadwallet` RPC command and some fixes.
- Removes the "Restart requirement" from wallet encryption
- Fixes issue #3521
- Backports bitcoin#13506 which was already backported in #3877 but the recently merged #4076 revert it.

### CODE REVIEW

File: src/wallet/db.cpp [hunk 1]
```diff
@@ -504,11 +520,11 @@ BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bo
             // versions of BDB have an set_lk_exclusive method for this
             // purpose, but the older version we use does not.)
             for (auto& env : g_dbenvs) {
-                CheckUniqueFileid(env.second, strFilename, *pdb_tempthis->env->m_fileids[strFilename]);
+                CheckUniqueFileid(env.second, strFilename, *pdb_temp, this->env->m_fileids[strFilename]);
```
Comment (UdjinM6) [inline src/wallet/db.cpp:523]: this fix belongs to 85349ca1f763cb5360edcea431f02331323be195

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3367,7 +3367,7 @@ static UniValue createwallet(const JSONRPCRequest& request)
 
     WalletLocation location(request.params[0].get_str());
     if (location.Exists()) {
-        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet " + wallet_name + " already exists.");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet " + location.GetName() + " already exists.");
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:3370]: this fix belongs to 8bb98a53373ddb20e6ec8b3a41c408b4e68c1d5d

File: src/wallet/test/coinjoin_tests.cpp [hunk 3]
```diff
@@ -36,7 +36,7 @@ class CTransactionBuilderTestSetup : public TestChain100Setup
     CTransactionBuilderTestSetup()
     {
         CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
-        wallet = MakeUnique<CWallet>("mock", WalletDatabase::CreateMock());
+        wallet = MakeUnique<CWallet>(WalletLocation(), WalletDatabase::CreateMock());
```
Comment (UdjinM6) [inline src/wallet/test/coinjoin_tests.cpp:39]: this fix belongs to 8bb98a53373ddb20e6ec8b3a41c408b4e68c1d5d

File: src/wallet/test/wallet_tests.cpp [hunk 4]
```diff
@@ -395,7 +395,7 @@ class CreateTransactionTestSetup : public TestChain100Setup
     CreateTransactionTestSetup()
     {
         CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
-        wallet = MakeUnique<CWallet>("mock", WalletDatabase::CreateMock());
+        wallet = MakeUnique<CWallet>(WalletLocation(), WalletDatabase::CreateMock());
```
Comment (UdjinM6) [inline src/wallet/test/wallet_tests.cpp:398]: this fix belongs to 8bb98a53373ddb20e6ec8b3a41c408b4e68c1d5d

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -4965,7 +4963,7 @@ bool CWallet::Verify(const WalletLocation& location, bool salvage_wallet, std::s
         return false;
     }
 
-    std::unique_ptr<CWallet> tempWallet = MakeUnique<CWallet>(location.GetName(), WalletDatabase::Create(wallet_path));
+    std::unique_ptr<CWallet> tempWallet = MakeUnique<CWallet>(location, WalletDatabase::Create(wallet_path));
```
Comment (UdjinM6) [inline src/wallet/wallet.cpp:4966]: this fix belongs to 8bb98a53373ddb20e6ec8b3a41c408b4e68c1d5d

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
pls see 77b5738eba, c3bff63839 and suggestions below

**Comment by xdustinface** [comment]
@UdjinM6 Ops 😄 moved the fixes (including yours) into the correct commits

**Review by thephez — APPROVED** [review]
Slightly tested ACK.
Confirmed #3521 fixed and `unloadwallet` appeared to work.

**Review by UdjinM6 — APPROVED** [review]
Slightly tested ACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Pull request merged**